lib/dry/types/builder.rb

```diff
@@ -30,8 +31,18 @@ def constructor_type
       #
       # @api private
       def |(other)
-        klass = constrained? && other.constrained? ? Sum::Constrained : Sum
-        klass.new(self, other)
+        compose(other, Sum)
+      end
+
+      # Compose two types into an Intersection type
+      #
+      # @param [Type] other
+      #
+      # @return [Intersection, Intersection::Constrained]
+      #
+      # @api private
+      def &(other)
+        compose(other, Intersection)
```

Comment:
This is actually a circular dependency! Any thoughts on how we could fix this issue?